File: automation.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <filesystem>
#include <sstream>
#include <iomanip>
#include <future>
#include <mutex>
#include <algorithm>
#include <thread>

namespace fs = std::filesystem;

/* ---------------- Structures ---------------- */

struct Task {
    std::string main_conf_file;       // filename inside temp folder
    std::string detector_conf_file;   // filename inside temp folder
    int run_id;
};

std::mutex cout_mutex;

/* ---------------- Helpers ---------------- */

std::string replace_parameter(const std::string& content,
                              const std::string& param,
                              const std::string& new_value) {
    std::stringstream ss(content);
    std::string line, result;
    bool replaced = false;

    while (std::getline(ss, line)) {
        size_t eq_pos = line.find('=');
        if (!replaced && eq_pos != std::string::npos) {
            std::string key = line.substr(0, eq_pos);
            key.erase(key.find_last_not_of(" \t") + 1);
            key.erase(0, key.find_first_not_of(" \t"));

            if (key == param) {
                result += param + " = " + new_value + "\n";
                replaced = true;
                continue;
            }
        }
        result += line + "\n";
    }
    return result;
}

/* ---------------- Worker ---------------- */

void run_simulation(const Task& task,
                    const fs::path& temp_dir,
                    const fs::path& output_dir) {

    fs::path project_root = fs::current_path();

    std::string command =
    "docker run -it --rm "
    "--user $(id -u):$(id -g) "
    "-w /project "
    "-v \"" + project_root.string() + ":/project\" "
    "apsq:g4-11.3.2-root-6.32 "
    "allpix -c /project/temp_configs/" + task.main_conf_file;

    int return_code = std::system(command.c_str());

    std::lock_guard<std::mutex> lock(cout_mutex);
    if (return_code != 0) {
        std::cerr << "Run " << task.run_id << " failed\n";
    } else {
        std::cout << "Run " << task.run_id << " completed\n";
    }
}




/* ---------------- Main ---------------- */

int main() {
    // Folders
    fs::path temp_dir   = fs::current_path() / "temp_configs";
    fs::path output_dir = fs::current_path() / "output";
    fs::create_directories(temp_dir);
    fs::create_directories(output_dir);

    auto read_file = [](const std::string& path) {
        std::ifstream f(path);
        return std::string((std::istreambuf_iterator<char>(f)),
                            std::istreambuf_iterator<char>());
    };

    std::string config_content   = read_file("spacepix3_main.conf");
    std::string detector_content = read_file("spacepix3_detector.conf");

    /* -------- Copy the detector model into temp once -------- */
    fs::path model_file = "spacepix3_model.conf";
    fs::path model_copy = temp_dir / model_file.filename();
    fs::copy_file(model_file, model_copy, fs::copy_options::overwrite_existing);
    /* ------------------------------------------------------------------------ */

    std::vector<std::string> p_types = {"proton", "e-"};
    std::vector<std::string> orientations = {
        "0deg 0deg 0deg",
        "15deg 0deg 0deg",
        "0deg 15deg 0deg"
    };

    std::vector<Task> tasks;
    int run_counter = 0;

    /* -------- Generate tasks and temp config files -------- */

    for (int i = 0; i <= 1; ++i) {
        double energy = 5.0 + i * 0.1;
        std::stringstream e_ss;
        e_ss << std::fixed << std::setprecision(1) << energy;
        std::string e_str = e_ss.str();
        std::string e_file = e_str;
        std::replace(e_file.begin(), e_file.end(), '.', '_');

        for (const auto& ptype : p_types) {
            for (const auto& orient : orientations) {
                std::string o_clean = orient;
                std::replace(o_clean.begin(), o_clean.end(), ' ', '_');

                // Filenames inside temp folder
                std::string det_name  = "detector_auto_" + e_file + "_" + ptype + "_" + o_clean + ".conf";
                std::string main_name = "main_auto_" + e_file + "_" + ptype + "_" + o_clean + ".conf";
                std::string out_path =
                    "temp_output/data_auto_" + e_file + "_" + ptype + "_" + o_clean + ".root";
                    //temp_output will be destroyed after making the data table

                // Replace parameters
                std::string run_det = replace_parameter(detector_content, "orientation", orient);
                std::string run_main = replace_parameter(config_content, "file_name", "\"" + out_path + "\"");
                run_main = replace_parameter(run_main, "source_energy", e_str + "GeV");
                run_main = replace_parameter(run_main, "particle_type", "\"" + ptype + "\"");
                run_main = replace_parameter(run_main, "detectors_file", "\"" + det_name + "\"");

                // Write temp files
                std::ofstream(temp_dir / det_name)  << run_det;
                std::ofstream(temp_dir / main_name) << run_main;

                tasks.push_back({main_name, det_name, run_counter++});
            }
        }
    }

    /* -------- Parallel execution (limited) -------- */

    const unsigned max_parallel =
        std::max(1u, std::thread::hardware_concurrency());
    std::cout << "Starting " << tasks.size()
              << " simulations with max "
              << max_parallel << " in parallel\n";

    std::vector<std::future<void>> futures;

    for (const auto& task : tasks) {
        while (futures.size() >= max_parallel) {
            futures.front().get();
            futures.erase(futures.begin());
        }
        futures.push_back(std::async(std::launch::async,
                                     run_simulation,
                                     task,
                                     temp_dir,
                                     output_dir));
    }

    for (auto& f : futures)
        f.get();
	
	
	// write root outputs to shared .root file with flattened tree
	// flattened tree can then be easily read with pythons uproot 
	fs::path project_root = fs::current_path();
	fs::path temp_output_dir = "/project/output/temp_output";
	std::string output_reading_command =
    "docker run --rm "
    "--user $(id -u):$(id -g) "
    "-w /project "
    "-v \"" + project_root.string() + ":/project\" "
    "apsq:g4-11.3.2-root-6.32 "
    "root -l -b -q "
    "-e '.L /opt/allpix/lib/libAllpixObjects.so' "
    "-e '.L OutputReader3.C++' "
    "-e 'treeMerge(\"" + temp_output_dir.string() + "\")'";

    
    int return_code = std::system(output_reading_command.c_str());

    std::lock_guard<std::mutex> lock(cout_mutex);
    if (return_code != 0) {
        std::cerr << "Output reading failed.\n";
    } else {
        std::cout << "Output reading succeeded.\n";
    }
	
	
	
    /* -------- Cleanup temp folders -------- */
    fs::remove_all(temp_dir);
    fs::remove_all("output/temp_output");

    std::cout << "All simulations processed. All temp folders deleted.\n";
    return 0;
}

File: automation.py
import os
import subprocess
from multiprocessing import Pool
try:
    from tqdm import tqdm
except ImportError:
    print("tqdm not found. Install it using 'pip install tqdm' for a progress bar.")
    tqdm = None

def replace_parameter(content, param, new_value):
    lines = content.splitlines()
    for i, line in enumerate(lines):
        if line.strip().startswith(param):
            lines[i] = f"{param} = {new_value}"
            return "\n".join(lines)
    return content

def read_file_content(filename):
    try:
        with open(filename, 'r') as f:
            return f.read()
    except FileNotFoundError:
        print(f"Error: Cannot open file {filename}!")
        exit(1)

def run_simulation(task):
    """
    Function executed by worker processes.
    Returns a status message upon completion.
    """
    main_conf = task['main_conf']
    det_conf = task['det_conf']
    run_id = task['run_id']

    command = [
        "docker", "run", "--rm",
        "-w", "/pulse_simulation",
        "-v", f"{os.getcwd()}:/pulse_simulation",
        "apsq:g4-11.3.2-root-6.32",
        "allpix", "-c", f"/pulse_simulation/{main_conf}", 
        "-d", f"/pulse_simulation/{det_conf}"
    ]

    # capture_output=True keeps the terminal clean during parallel execution
    result = subprocess.run(command, capture_output=True, text=True)
    
    if result.returncode != 0:
        return f"Failure in Run {run_id} ({main_conf})"
    return f"Success: Run {run_id}"

def main():
    os.makedirs("output_auto", exist_ok=True)

    template_file = "spacepix3_main.conf"
    detector_file = "spacepix3_detector.conf"

    config_content = read_file_content(template_file)
    detector_content = read_file_content(detector_file)

    # Simulation matrix
    energies = [round(5.0 + i * 0.1, 2) for i in range(int((10.0 - 5.0) / 0.1) + 1)]
    particle_types = ["proton", "e-"]
    orientations = ["0deg 0deg 0deg", "15deg 0deg 0deg", "0deg 15deg 0deg"]

    tasks = []
    run_counter = 0

    # 1. Preparation: Create configs
    for energy in energies:
        for ptype in particle_types:
            for orientation in orientations:
                energy_str = str(energy).replace('.', '_')
                orient_clean = orientation.replace(' ', '_')
                
                det_name = f"detector_auto_{energy_str}_{ptype}_{orient_clean}.conf"
                main_name = f"main_auto_{energy_str}_{ptype}_{orient_clean}.conf"
                output_file = f"/pulse_simulation/output_auto/data_auto_{energy_str}_{ptype}_{orient_clean}.root"

                run_main = replace_parameter(config_content, "file_name", f'"{output_file}"')
                run_main = replace_parameter(run_main, "source_energy", f"{energy}GeV")
                run_main = replace_parameter(run_main, "particle_type", f'"{ptype}"')
                run_det = replace_parameter(detector_content, "orientation", orientation)

                with open(det_name, 'w') as f: f.write(run_det)
                with open(main_name, 'w') as f: f.write(run_main)

                tasks.append({'main_conf': main_name, 'det_conf': det_name, 'run_id': run_counter})
                run_counter += 1

    # 2. Execution: Parallel Pool with TQDM
    print(f"Launching {len(tasks)} simulations...")
    
    results = []
    with Pool() as pool:
        # Use imap or imap_unordered to work with tqdm
        if tqdm:
            # Wrap the iterator with tqdm for the progress bar
            for result in tqdm(pool.imap_unordered(run_simulation, tasks), total=len(tasks), desc="Simulating"):
                results.append(result)
        else:
            results = pool.map(run_simulation, tasks)

    # 3. Final Summary
    failures = [res for res in results if "Failure" in res]
    print(f"\n--- Processing Complete ---")
    print(f"Total runs: {len(results)}")
    print(f"Failures: {len(failures)}")
    for fail in failures:
        print(fail)

if __name__ == "__main__":
    main()